Samples/WMCConsole_winrtcpp/WMCConsole_winrtcpp.cpp

```diff
@@ -31,115 +31,247 @@ using namespace Windows::Storage;
+            int frameRateDec = ((int)(format.FrameRate() * 100) % 100);
+            idx++;
+            auto formatStr = std::to_wstring(idx) + L". " + std::wstring(format.Subtype().c_str()) + L": " +
+                std::to_wstring(format.Width()) + L"x" + std::to_wstring(format.Height()) + L" @ " + std::to_wstring(frameRateInt) + L"." + std::to_wstring(frameRateDec) + L" fps";
+            std::wcout << formatStr << std::endl;
+            mediaDescriptionIter.MoveNext();
+        }
+        int selection;
+        do
+        {
+            std::wcout << L"Enter selection:";
+            std::cin >> selection;
+        } while ((selection > idx) || (selection < 1));
+        return selection - 1; // Because selection idx is from 1 to N and indexes for from 0 to N-1
+    }
+
+    //
+    // Create a Frame reader for Mock preview and drain samples from preview/record pin for auto-exposure conververgence
+    //
+    void StartMockPreview(MediaCapture mediaCapture)
+    {
+        // start Preview to get the 3A running and wait for convergence.
+        MediaFrameSource previewframeSource(nullptr);
+        MediaStreamType streamTypelookup = MediaStreamType::VideoPreview;
+
+        // Try to find the suitable pin where 3A will be running.
+        // Start by looking for a preview pin , if not found look for record pin 
+        // However exit the loop when preview and record pins are not available as just running the photo pin cannot converge 3A
+        while ((previewframeSource == nullptr) && (streamTypelookup != MediaStreamType::Photo))
         {
-            panelLocation = L"\\" + panelTypes[enclosureLocation.Panel()];
+            auto frameSourceIter = mediaCapture.FrameSources().First();
+            // If there is no preview pin, find a record pin
+            while (frameSourceIter.HasCurrent())
+            {
+                auto frameSource = frameSourceIter.Current().Value();
+                if (frameSource.Info().MediaStreamType() == streamTypelookup)
+                {
+                    previewframeSource = frameSource;
+                    break;
+                }
+                frameSourceIter.MoveNext();
+            }
+            streamTypelookup = (streamTypelookup == MediaStreamType::VideoPreview) ? MediaStreamType::VideoRecord : MediaStreamType::Photo;
         }
 
-        std::wcout << idx << L":" << currSource.SourceGroup().DisplayName().c_str() <<"-->" << currSource.DeviceInformation().Name().c_str() << panelLocation.c_str() << ":" << streamTypes[currSource.MediaStreamType()].c_str() << std::endl;
-        source.MoveNext();
+        m_PreviewFrameReader = mediaCapture.CreateFrameReaderAsync(previewframeSource).get();
+        m_PreviewFrameReader.AcquisitionMode(MediaFrameReaderAcquisitionMode::Realtime);
+        m_PreviewFrameReader.StartAsync().get();
+        std::cout << "\nDraining samples for one second from the pipeline to enable 3A convergence\n";
+        Sleep(1000);

```

Comment:
>Sleep(1000) [](start = 8, length = 11)

suggestion: you can substitue with "co_await 1s;"for a C++/WinRT way of awaiting passively

Reply:
but the idea here is to block the further execution and not take photos for 1sec. With co_await eventually up the stack I need to put this in a Task and as I understand co_await will only keep this Task call stack active for 1s and proceed execution of the rest of the routines, unless I wait, which is the same effect. The rest of the sample does not use co_routines directly to keep things simple.

---
In reply to: [286776586](https://github.com/microsoft/Windows-Camera/pull/10#discussion_r286776586) [](ancestors = 286776586)